### PR TITLE
feat: Add Api Authentication security integration to sdk

### PR DESCRIPTION
**Pull request by sfc-gh-jmichalak**
Add Api Authentication security integration to sdk. Implement all 3 flows, but integration tests for JWT bearer flow are skipped due to errors during create. This will be fixed in a follow up.

## Test Plan

* [x] unit tests

* [x] integration tests

## References

https://docs.snowflake.com/en/sql-reference/sql/create-security-integration-api-auth
https://docs.snowflake.com/en/sql-reference/sql/alter-security-integration-api-auth

### CODE REVIEW

File: pkg/sdk/testint/tasks_gen_integration_test.go [hunk 1]
```diff
@@ -531,7 +531,7 @@ func TestInt_Tasks(t *testing.T) {
 		returnedTasks, err := client.Tasks.Show(ctx, showRequest)
 		require.NoError(t, err)
 
-		assert.Equal(t, 2, len(returnedTasks))
+		assert.LessOrEqual(t, 2, len(returnedTasks))
```
Comment (sfc-gh-jcieslak) [inline pkg/sdk/testint/tasks_gen_integration_test.go:534]: Why `LessOrEqual`? I would expect this test to fail with a number greater than two 🤔 (same in views)

File: pkg/sdk/security_integrations_def.go [hunk 2]
```diff
@@ -246,6 +342,86 @@ var SecurityIntegrationsDef = g.NewInterface(
...
+				).
+				TextAssignment("OAUTH_CLIENT_ID", g.ParameterOptions().Required().SingleQuotes()).
+				TextAssignment("OAUTH_CLIENT_SECRET", g.ParameterOptions().Required().SingleQuotes()).
+				OptionalAssignment(
+					"OAUTH_GRANT",
+					g.KindOfT[ApiAuthenticationSecurityIntegrationOauthGrantOption](),
+					g.ParameterOptions(),
+				).
+				OptionalNumberAssignment("OAUTH_ACCESS_TOKEN_VALIDITY", g.ParameterOptions()).
+				OptionalNumberAssignment("OAUTH_REFRESH_TOKEN_VALIDITY", g.ParameterOptions()).
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/security_integrations_def.go:367]: +1

File: pkg/sdk/security_integrations_def.go [hunk 3]
```diff
@@ -246,6 +342,86 @@ var SecurityIntegrationsDef = g.NewInterface(
...
+	).
+	CustomOperation(
+		"CreateApiAuthenticationJwtBearerFlow",
+		"https://docs.snowflake.com/en/sql-reference/sql/create-security-integration-api-auth",
+		createSecurityIntegrationOperation("CreateApiAuthenticationJwtBearerFlow", func(qs *g.QueryStruct) *g.QueryStruct {
+			return qs.
+				PredefinedQueryStructField("integrationType", "string", g.StaticOptions().SQL("TYPE = API_AUTHENTICATION")).
+				PredefinedQueryStructField("authType", "string", g.StaticOptions().SQL("AUTH_TYPE = OAUTH2")).
+				BooleanAssignment("ENABLED", g.ParameterOptions().Required()).
+				TextAssignment("OAUTH_ASSERTION_ISSUER", g.ParameterOptions().Required().SingleQuotes()).
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/security_integrations_def.go:406]: +1

File: pkg/sdk/security_integrations_def.go [hunk 4]
```diff
@@ -109,6 +124,87 @@ func alterSecurityIntegrationOperation(structName string, opts func(qs *g.QueryS
...
+	OptionalTextAssignment("OAUTH_TOKEN_ENDPOINT", g.ParameterOptions().SingleQuotes()).
+	OptionalAssignment(
+		"OAUTH_CLIENT_AUTH_METHOD",
+		g.KindOfT[ApiAuthenticationSecurityIntegrationOauthClientAuthMethodOption](),
+		g.ParameterOptions(),
+	).
+	OptionalTextAssignment("OAUTH_CLIENT_ID", g.ParameterOptions().SingleQuotes()).
+	OptionalTextAssignment("OAUTH_CLIENT_SECRET", g.ParameterOptions().SingleQuotes()).
+	OptionalAssignment(
+		"OAUTH_GRANT",
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/security_integrations_def.go:138]: similar comment as to create

File: pkg/sdk/security_integrations_def.go [hunk 5]
```diff
@@ -109,6 +124,87 @@ func alterSecurityIntegrationOperation(structName string, opts func(qs *g.QueryS
...
+		g.KindOfT[ApiAuthenticationSecurityIntegrationOauthGrantOption](),
+		g.ParameterOptions(),
+	).
+	OptionalNumberAssignment("OAUTH_ACCESS_TOKEN_VALIDITY", g.ParameterOptions()).
+	OptionalNumberAssignment("OAUTH_REFRESH_TOKEN_VALIDITY", g.ParameterOptions()).
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/security_integrations_def.go:143]: similar comment as to create

File: pkg/sdk/security_integrations_def.go [hunk 6]
```diff
@@ -109,6 +124,87 @@ func alterSecurityIntegrationOperation(structName string, opts func(qs *g.QueryS
...
+	OptionalComment().
+	WithValidation(g.AtLeastOneValueSet, "Enabled", "OauthAuthorizationEndpoint", "OauthTokenEndpoint", "OauthClientAuthMethod", "OauthClientId", "OauthClientSecret", "OauthGrant",
+		"OauthAccessTokenValidity", "Comment")
+
+var apiAuthJwtBearerFlowIntegrationUnsetDef = g.NewQueryStruct("ApiAuthenticationJwtBearerFlowIntegrationUnset").
+	OptionalSQL("ENABLED").
+	OptionalSQL("COMMENT").
+	WithValidation(g.AtLeastOneValueSet, "Enabled", "Comment")
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/security_integrations_def.go:206]: we cannot unset any other thing? (I know not visible in the docs but have you tested?)

File: pkg/sdk/testint/tasks_gen_integration_test.go (hunk 1 above)
Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/tasks_gen_integration_test.go:534]: xD

Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/tasks_gen_integration_test.go:534]: Golang

File: pkg/sdk/testint/materialized_views_gen_integration_test.go [hunk 7]
```diff
@@ -360,7 +360,8 @@ func TestInt_MaterializedViews(t *testing.T) {
 
 	t.Run("show materialized view: no existing view", func(t *testing.T) {
 		showRequest := sdk.NewShowMaterializedViewRequest().
-			WithIn(&sdk.In{Schema: sdk.NewDatabaseObjectIdentifier(testDb(t).Name, testSchema(t).Name)})
+			WithIn(&sdk.In{Schema: sdk.NewDatabaseObjectIdentifier(testDb(t).Name, testSchema(t).Name)}).
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/materialized_views_gen_integration_test.go:363]: nope.

Please use `testClientHelper().Ids.SchemaId()`

File: pkg/sdk/testint/security_integrations_gen_integration_test.go [hunk 8]
```diff
@@ -429,26 +595,267 @@ func TestInt_SecurityIntegrations(t *testing.T) {
...
+		}
+		alterRequestUnsetTags := sdk.NewAlterApiAuthenticationAuthorizationCodeGrantFlowSecurityIntegrationRequest(id).WithUnsetTags(unsetTags)
+
+		err = client.SecurityIntegrations.AlterApiAuthenticationAuthorizationCodeGrantFlow(ctx, alterRequestUnsetTags)
+		require.NoError(t, err)
+
+		_, err = client.SystemFunctions.GetTag(ctx, tag.ID(), id, sdk.ObjectTypeIntegration)
+		require.Error(t, err)
+	})
+	t.Run("AlterApiAuthenticationJwtBearerFlow", func(t *testing.T) {
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/security_integrations_gen_integration_test.go:769]: nit-pick: we use interlines between the tests

File: pkg/sdk/security_integrations_gen_test.go [hunk 9]
```diff
@@ -4,6 +4,149 @@ import (
...
+		opts.IfNotExists = Bool(true)
+		opts.OauthTokenEndpoint = Pointer("foo")
+		opts.OauthClientAuthMethod = Pointer(ApiAuthenticationSecurityIntegrationOauthClientAuthMethodClientSecretPost)
+		opts.OauthGrant = Pointer(ApiAuthenticationSecurityIntegrationOauthClientGrantClientCredentials)
+		opts.OauthAccessTokenValidity = Pointer(42)
+		opts.OauthRefreshTokenValidity = Pointer(42)
+		opts.OauthAllowedScopes = []AllowedScope{{Scope: "bar"}}
+		opts.Comment = Pointer("foo")
+		assertOptsValidAndSQLEquals(t, opts, "CREATE SECURITY INTEGRATION IF NOT EXISTS %s TYPE = API_AUTHENTICATION AUTH_TYPE = OAUTH2 ENABLED = true OAUTH_TOKEN_ENDPOINT = 'foo'"+
+			" OAUTH_CLIENT_AUTH_METHOD = CLIENT_SECRET_POST OAUTH_CLIENT_ID = 'foo' OAUTH_CLIENT_SECRET = 'bar' OAUTH_GRANT = CLIENT_CREDENTIALS"+
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/security_integrations_gen_test.go:50]: +1

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: pkg/sdk/security_integrations_def.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_def.go:367]: Right, it was reported in DOC-2249.

File: pkg/sdk/security_integrations_def.go [hunk 10]
```diff
@@ -109,6 +124,87 @@ func alterSecurityIntegrationOperation(structName string, opts func(qs *g.QueryS
...
+		g.ParameterOptions(),
+	).
+	OptionalNumberAssignment("OAUTH_ACCESS_TOKEN_VALIDITY", g.ParameterOptions()).
+	OptionalNumberAssignment("OAUTH_REFRESH_TOKEN_VALIDITY", g.ParameterOptions()).
+	ListAssignment("OAUTH_ALLOWED_SCOPES", "AllowedScope", g.ParameterOptions().Parentheses()).
+	OptionalComment().
+	WithValidation(g.AtLeastOneValueSet, "Enabled", "OauthTokenEndpoint", "OauthClientAuthMethod", "OauthClientId", "OauthClientSecret", "OauthGrant",
+		"OauthAccessTokenValidity", "OauthRefreshTokenValidity", "OauthAllowedScopes", "Comment")
+
+var apiAuthClientCredentialsFlowIntegrationUnsetDef = g.NewQueryStruct("ApiAuthenticationClientCredentialsFlowIntegrationUnset").
```
[the remark this answers is not in the document]
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_def.go:149]: Already reported in DOC-2249

File: pkg/sdk/testint/tasks_gen_integration_test.go (hunk 1 above)
Comment (sfc-gh-jmichalak) [reply pkg/sdk/testint/tasks_gen_integration_test.go:534]: During some tests executed in parallel, Show returns more objects and that's why we should assert for at least 2 objects. We're still checking objects details below.

File: pkg/sdk/security_integrations_gen.go [hunk 11]
```diff
@@ -23,6 +29,74 @@ type SecurityIntegrations interface {
...
+	create                     bool                                                             `ddl:"static" sql:"CREATE"`
+	OrReplace                  *bool                                                            `ddl:"keyword" sql:"OR REPLACE"`
+	securityIntegration        bool                                                             `ddl:"static" sql:"SECURITY INTEGRATION"`
+	IfNotExists                *bool                                                            `ddl:"keyword" sql:"IF NOT EXISTS"`
+	name                       AccountObjectIdentifier                                          `ddl:"identifier"`
+	integrationType            string                                                           `ddl:"static" sql:"TYPE = API_AUTHENTICATION"`
+	authType                   string                                                           `ddl:"static" sql:"AUTH_TYPE = OAUTH2"`
+	Enabled                    bool                                                             `ddl:"parameter" sql:"ENABLED"`
+	OauthAssertionIssuer       string                                                           `ddl:"parameter,single_quotes" sql:"OAUTH_ASSERTION_ISSUER"`
```
[the remark this answers is not in the document]
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_gen.go:88]: Not only this field is possible, but it's also required for JWT flow xD I reported this in DOC-2249.

File: pkg/sdk/security_integrations_def.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_def.go:406]: This field is actually required in JWT flow. I reported this in DOC-2249.

File: pkg/sdk/security_integrations_gen_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_gen_test.go:50]: Good catch. It works without single quotes, I added this to DOC-2249.

File: pkg/sdk/security_integrations_def.go (hunk 6 above)
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_def.go:206]: For other flows I tested each field and it looks like only `ENABLED` and `COMMENT` can be unset. Jwt flow was not properly tested due to problems with `OAUTH_CLIENT_SECRET` during create, we discussed about this.

File: pkg/sdk/security_integrations_def.go [hunk 12]
```diff
@@ -246,6 +342,86 @@ var SecurityIntegrationsDef = g.NewInterface(
 	"SecurityIntegration",
 	g.KindOfT[AccountObjectIdentifier](),
 ).
+	CustomOperation(
+		"CreateApiAuthenticationClientCredentialsFlow",
```
[the remark this answers is not in the document]
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_def.go]: Renamed (added `With`).

File: pkg/sdk/security_integrations_def.go [hunk 13]
```diff
@@ -246,6 +342,86 @@ var SecurityIntegrationsDef = g.NewInterface(
...
+		"CreateApiAuthenticationAuthorizationCodeGrantFlow",
+		"https://docs.snowflake.com/en/sql-reference/sql/create-security-integration-api-auth",
+		createSecurityIntegrationOperation("CreateApiAuthenticationAuthorizationCodeGrantFlow", func(qs *g.QueryStruct) *g.QueryStruct {
+			return qs.
+				PredefinedQueryStructField("integrationType", "string", g.StaticOptions().SQL("TYPE = API_AUTHENTICATION")).
+				PredefinedQueryStructField("authType", "string", g.StaticOptions().SQL("AUTH_TYPE = OAUTH2")).
+				BooleanAssignment("ENABLED", g.ParameterOptions().Required()).
+				OptionalTextAssignment("OAUTH_AUTHORIZATION_ENDPOINT", g.ParameterOptions().SingleQuotes()).
+				OptionalTextAssignment("OAUTH_TOKEN_ENDPOINT", g.ParameterOptions().SingleQuotes()).
+				OptionalAssignment(
```
[the remark this answers is not in the document]
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_def.go:358]: Fixed.

File: pkg/sdk/security_integrations_def.go (hunk 4 above)
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_def.go]: Fixed.

File: pkg/sdk/security_integrations_def.go (hunk 5 above)
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_def.go:131]: Fixed.

File: pkg/sdk/testint/materialized_views_gen_integration_test.go (hunk 7 above)
Comment (sfc-gh-jmichalak) [reply pkg/sdk/testint/materialized_views_gen_integration_test.go]: Done.

File: pkg/sdk/testint/security_integrations_gen_integration_test.go (hunk 8 above)
Comment (sfc-gh-jmichalak) [reply pkg/sdk/testint/security_integrations_gen_integration_test.go]: Fixed.

File: pkg/sdk/security_integrations_def.go [hunk 14]
```diff
@@ -246,6 +342,86 @@ var SecurityIntegrationsDef = g.NewInterface(
...
+				BooleanAssignment("ENABLED", g.ParameterOptions().Required()).
+				OptionalTextAssignment("OAUTH_TOKEN_ENDPOINT", g.ParameterOptions().SingleQuotes()).
+				OptionalAssignment(
+					"OAUTH_CLIENT_AUTH_METHOD",
+					g.KindOfT[ApiAuthenticationSecurityIntegrationOauthClientAuthMethodOption](),
+					g.ParameterOptions(),
+				).
+				TextAssignment("OAUTH_CLIENT_ID", g.ParameterOptions().Required().SingleQuotes()).
+				TextAssignment("OAUTH_CLIENT_SECRET", g.ParameterOptions().Required().SingleQuotes()).
+				OptionalAssignment(
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/security_integrations_def.go]: I am still curious if `(...)adding it/omitting it change[s] anything?`

[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/security_integrations_def.go]: Because if there is no difference, let's either:
- skip it
- or make it always present

**Review by sfc-gh-asawicki — DISMISSED** [review]

**Review by sfc-gh-jcieslak — APPROVED** [review]

File: pkg/sdk/security_integrations_def.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jmichalak) [reply pkg/sdk/security_integrations_def.go]: Well, we're checking `OAUTH_GRANT` value from describe. If it's omitted in a query, this field is empty.

**Review by sfc-gh-asawicki — APPROVED** [review]

**Pull request merged**